File: apps/api-rs/src/agent/tools/clarification.rs
use anyhow::{bail, Result};
use serde::Deserialize;

use super::{AgentTool, TerminalToolEffect, ToolEffect, ToolExecution, ToolExecutionContext};
use crate::agent::model::{AgentToolCall, AgentToolDefinition};
use crate::models::agent::AgentMode;
use crate::models::{Confidence, NoAnswerReason};

pub struct ClarificationTool;

#[derive(Debug, Deserialize)]
struct ClarificationArguments {
    question: String,
    reason: String,
}

#[async_trait::async_trait]
impl AgentTool for ClarificationTool {
    fn definition(&self) -> AgentToolDefinition {
        AgentToolDefinition {
            name: "ask_clarification".to_string(),
            description: "Pause and ask one precise question when the user's intent is genuinely ambiguous. Do not use this for missing evidence or uncertain corpus contents.".to_string(),
            parameters: serde_json::json!({
                "type": "object",
                "additionalProperties": false,
                "properties": {
                    "question": {
                        "type": "string",
                        "description": "One concise user-facing clarification question."
                    },
                    "reason": {
                        "type": "string",
                        "description": "Brief operational reason; do not expose hidden reasoning."
                    }
                },
                "required": ["question", "reason"]
            }),
        }
    }

    async fn execute(
        &self,
        call: &AgentToolCall,
        _context: &ToolExecutionContext<'_>,
    ) -> Result<ToolExecution> {
        let arguments: ClarificationArguments = serde_json::from_str(&call.arguments_json)?;
        let question = arguments.question.trim().to_string();
        if question.is_empty() {
            bail!("ask_clarification requires a non-empty question");
        }
        Ok(ToolExecution {
            public_result: serde_json::json!({
                "question": question,
                "reason": arguments.reason,
                "status": "waiting_for_user"
            }),
            model_result: serde_json::json!({
                "status": "waiting_for_user",
                "question": question
            }),
            effect: ToolEffect::Terminal(TerminalToolEffect {
                answer: question,
                mode: AgentMode::Clarifier,
                confidence: Confidence::Low,
                no_answer_reason: Some(NoAnswerReason::NeedsClarification),
            }),
        })
    }

    fn component_name(&self) -> String {
        "builtin-ask-clarification-v1".to_string()
    }
}

File: apps/api-bun/src/agent/tools/clarification.ts
// 移植自 apps/api-rs/src/agent/tools/clarification.rs
import type { AgentToolCall, AgentToolDefinition } from '../model.ts';
import type { AgentTool, AgentToolContext, ToolExecution } from './types.ts';

interface ClarificationArguments {
  question: string;
  reason: string;
}

function parseClarificationArguments(argumentsJson: string): ClarificationArguments {
  let value: unknown;
  try {
    value = JSON.parse(argumentsJson);
  } catch (error) {
    throw new Error(`invalid clarification arguments: ${(error as Error).message}`);
  }
  if (typeof value !== 'object' || value === null) {
    throw new Error('invalid clarification arguments: expected an object');
  }
  const record = value as Record<string, unknown>;
  if (typeof record.question !== 'string') {
    throw new Error('invalid clarification arguments: missing field \`question\`');
  }
  if (typeof record.reason !== 'string') {
    throw new Error('invalid clarification arguments: missing field \`reason\`');
  }
  return { question: record.question, reason: record.reason };
}

export class ClarificationTool implements AgentTool {
  definition(): AgentToolDefinition {
    return {
      name: 'ask_clarification',
      description: 'Pause and ask one precise question when the user\'s intent is genuinely ambiguous. Do not use this for missing evidence or uncertain corpus contents.',
      parameters: {
        type: 'object',
        additionalProperties: false,
        properties: {
          question: {
            type: 'string',
            description: 'One concise user-facing clarification question.',
          },
          reason: {
            type: 'string',
            description: 'Brief operational reason; do not expose hidden reasoning.',
          },
        },
        required: ['question', 'reason'],
      },
    };
  }

  async execute(call: AgentToolCall, _context: AgentToolContext): Promise<ToolExecution> {
    const arguments_ = parseClarificationArguments(call.arguments_json);
    const question = arguments_.question.trim();
    if (question.length === 0) {
      throw new Error('ask_clarification requires a non-empty question');
    }
    return {
      public_result: {
        question: question,
        reason: arguments_.reason,
        status: 'waiting_for_user',
      },
      model_result: {
        status: 'waiting_for_user',
        question: question,
      },
      effect: {
        type: 'terminal',
        terminal: {
          answer: question,
          mode: 'clarifier',
          confidence: 'low',
          no_answer_reason: 'needs_clarification',
        },
      },
    };
  }

  componentName(): string {
    return 'builtin-ask-clarification-v1';
  }
}
